ICE: stack overflow in diagnostics with 2 identical broken fns

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
fn a() -> _ {
    &a
}

fn a() -> _ {
    &a
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.71.0-nightly (fec9adcdb 2023-04-21)
binary: rustc
commit-hash: fec9adcdbc21469ef105162cc8cabf81c72d06be
commit-date: 2023-04-21
host: x86_64-unknown-linux-gnu
release: 1.71.0-nightly
LLVM version: 16.0.2
```

### Error output

```
error[E0428]: the name `a` is defined multiple times
 --> treereduce.out:5:1
  |
1 | fn a() -> _ {
  | ----------- previous definition of the value `a` here
...
5 | fn a() -> _ {
  | ^^^^^^^^^^^ `a` redefined here
  |
  = note: `a` must be defined only once in the value namespace of this module

error[E0601]: `main` function not found in crate `treereduce`
 --> treereduce.out:7:2
  |
7 | }
  |  ^ consider adding a `main` function to `treereduce.out`

error[E0121]: the placeholder `_` is not allowed within types on item signatures for return types
 --> treereduce.out:1:11
  |
1 | fn a() -> _ {
  |           ^ not allowed in type signatures
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x3207523)[0x7ff240a07523]
/usr/lib/libc.so.6(+0x38f50)[0x7ff23d4c1f50]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37b59e5)[0x7ff240fb59e5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x1f2)[0x7ff240fc4542]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x37865fc)[0x7ff240f865fc]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4e2d)[0x7ff240ee4e2d]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x36e4dac)[0x7ff240ee4dac]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(_RNvXs4_NtNtCslHyhO5NmBss_12rustc_middle2ty11diagnosticsNtB5_21MakeSuggestableFolderINtNtCs7aKEBUaXzWY_13rustc_type_ir4fold18FallibleTypeFolderNtNtB7_7context6TyCtxtE11try_fold_ty+0x175)[0x7ff240fc44c5]
/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e9ace2a97781a33e.so(+0x374f0c6)[0x7ff240f4f0c6]
.....
```

</p>
</details>

--- a/compiler/rustc_hir_analysis/src/collect.rs
+++ b/compiler/rustc_hir_analysis/src/collect.rs
@@ -1146,8 +1146,14 @@ fn infer_return_ty_for_fn_sig<'tcx>(
 
             let mut visitor = HirPlaceholderCollector::default();
             visitor.visit_ty(ty);
+
             let mut diag = bad_placeholder(tcx, visitor.0, "return type");
             let ret_ty = fn_sig.output();
+            // Don't leak types into signatures unless they're nameable!
+            // For example, if a function returns itself, we don't want that
+            // recursive function definition to leak out into the fn sig.
+            let mut should_recover = false;
+
             if let Some(ret_ty) = ret_ty.make_suggestable(tcx, false) {
                 diag.span_suggestion(
                     ty.span,
@@ -1155,15 +1161,7 @@ fn infer_return_ty_for_fn_sig<'tcx>(
                     ret_ty,
                     Applicability::MachineApplicable,
                 );
-            } else if matches!(ret_ty.kind(), ty::FnDef(..))
-                && let Some(fn_sig) = ret_ty.fn_sig(tcx).make_suggestable(tcx, false)
-            {
-                diag.span_suggestion(
-                    ty.span,
-                    "replace with the correct return type",
-                    fn_sig,
-                    Applicability::MachineApplicable,
-                );
+                should_recover = true;
             } else if let Some(sugg) = suggest_impl_trait(tcx, ret_ty, ty.span, def_id) {
                 diag.span_suggestion(
                     ty.span,
@@ -1181,9 +1179,20 @@ fn infer_return_ty_for_fn_sig<'tcx>(
                      https://doc.rust-lang.org/book/ch13-01-closures.html",
                 );
             }
-            diag.emit();
 
-            ty::Binder::dummy(fn_sig)
+            let guar = diag.emit();
+
+            if should_recover {
+                ty::Binder::dummy(fn_sig)
+            } else {
+                ty::Binder::dummy(tcx.mk_fn_sig(
+                    fn_sig.inputs().iter().copied(),
+                    tcx.ty_error(guar),
+                    fn_sig.c_variadic,
+                    fn_sig.unsafety,
+                    fn_sig.abi,
+                ))
+            }
         }
         None => icx.astconv().ty_of_fn(
             hir_id,

--- a/tests/ui/typeck/bad-recursive-type-sig-infer.rs
+++ b/tests/ui/typeck/bad-recursive-type-sig-infer.rs
@@ -1,0 +1,11 @@
+fn a() -> _ {
+    //~^ ERROR the placeholder `_` is not allowed within types on item signatures for return types
+    &a
+}
+
+fn b() -> _ {
+    //~^ ERROR the placeholder `_` is not allowed within types on item signatures for return types
+    &a
+}
+
+fn main() {}

--- a/tests/ui/typeck/bad-recursive-type-sig-infer.stderr
+++ b/tests/ui/typeck/bad-recursive-type-sig-infer.stderr
@@ -1,0 +1,15 @@
+error[E0121]: the placeholder `_` is not allowed within types on item signatures for return types
+  --> $DIR/bad-recursive-type-sig-infer.rs:1:11
+   |
+LL | fn a() -> _ {
+   |           ^ not allowed in type signatures
+
+error[E0121]: the placeholder `_` is not allowed within types on item signatures for return types
+  --> $DIR/bad-recursive-type-sig-infer.rs:6:11
+   |
+LL | fn b() -> _ {
+   |           ^ not allowed in type signatures
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0121`.

